### ISSUE TITLE
Add support for making an ArrayPool with pre-pinned buffers

### ISSUE THREAD

**Issue by davidfowl** [dotnet/coreclr#19936]
For IO scenarios, the memory pool is useful to save repeated allocations for socket reads and writes but there's no easy way to avoid pinning and unpinned of those `byte[]`. The other issue is long running operations that pin buffers that may end up causing GC fragmentation. We want to be able to say that these buffers will be used for IO and should be pinned until they go away. 

With that, we maybe able to remove the memory pool from Kestrel completely (I'm taking a look into that right now).

cc @jkotas

**Comment by jkotas** [comment]
@Maoni0 Do you think it would make sense to have an API to allocate pre-pinned memory, similar to the API to allocate memory in Gen2?

**Comment by Maoni0** [comment]
@davidfowl are you in control of allocating these buffers?

**Comment by davidfowl** [comment]
@Maoni0 Yes, today we're using this memory pool https://github.com/aspnet/KestrelHttpServer/blob/55e5e564226e7b27742c91f46ea71841071e4ac3/shared/Microsoft.Extensions.Buffers.MemoryPool.Sources/SlabMemoryPool.cs.

**Comment by Maoni0** [comment]
@davidfowl  ok that's great; I contemplated a few times in the past to provide a pinned heap but I wasn't seeing enough cases that warranted it as mostly folks who worked on higher level code that often did not have control over the allocation of objects they pinned. if you are going to do both the allocating and pinning it would make sense to provide this special heap. 

@jkotas yes, I will update the API proposal.

**Comment by geoffkizer** [comment]
Yes, a pinned heap definitely helps for buffer scenarios. We have a bunch of these for things like SocketsHttpHandler and SslStream.
